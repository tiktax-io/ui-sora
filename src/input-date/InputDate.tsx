import React, { ForwardedRef, forwardRef, HTMLProps, memo } from 'react'
import Icon from '../icon/Icon'
import { IconProps } from '../icon/Icon.types'
import { danger, success } from '../_styles'

import { isInvalidStyle, isValidStyle, useStyles } from './InputDate.styles'
import { InputDateProps } from './InputDate.types'

/******************************************************************************
 * Functional component that renders an input of type "date" element.
 * @param {CSSProperties} css CSS properties to add to the div wrapping the
 * element.
 * @param {string} dataTestId id of the container element to be
 * used on tests as a way to query the component.
 * @param {string} feedbackMessage validation message to display below the
 * input field.
 * @param {boolean} isInvalid true if the input will display an error status.
 * @param {boolean} isValid true if the input will display a success status.
 * @param {string} label label title of the input.
 * @property {string} min oldest date available in range: 'YYYY-MM-DD'.
 * @property {string} max oldest date available in rangE: 'YYYY-MM-DD'.
 * the selected option has been updated.
 * @property {number} step integer number representing days.
 * @param {(value: string) => unknown | void} onChange action to trigger when
 * @param {string} value the default option selected on render.
 * @returns {ReactElement} Input of type select.
 *****************************************************************************/
const InputDate = memo(forwardRef(({
  css,
  dataTestId,
  feedbackMessage,
  isInvalid,
  isValid,
  label,
  min,
  max,
  onChange,
  step,
  value,
}: InputDateProps, ref: ForwardedRef<HTMLInputElement>) => {
  const classes = useStyles()

  const feedbackIconProps: IconProps = {
    color: isInvalid ? danger : success,
    height: 16,
    icon: isInvalid ? 'danger' : 'success',
    width: 16
  }

  /****************************************************************************
   * HTML attributes and props to be passed to input box.
   ***************************************************************************/
  const inputProps: HTMLProps<HTMLInputElement> = {
    className: classes.inputBox,
    placeholder: 'hey',
    ref: ref,
    style: {
      ...isInvalid ? isInvalidStyle : isValid ? isValidStyle : null,
      ...css
    },
    type: 'date'
  }

  const spanFeedbackMessageAttributes = {
    className: classes.feedbackMessageContainer,
    'data-testid': `${dataTestId}SpanFeedbackMessage`,
    style: {
      color: isInvalid ? danger : success
    }
  }

  return (
    <div className={classes.container}>
      {label && <label className={classes.label}>{label}</label>}
      <input {...inputProps} />
      {(isInvalid || isValid) && <span {...spanFeedbackMessageAttributes}>
        <Icon {...feedbackIconProps} />
        <span className={classes.feedbackMessageText}>{feedbackMessage}</span>
      </span>}
    </div>
  )
}))

export default InputDate
