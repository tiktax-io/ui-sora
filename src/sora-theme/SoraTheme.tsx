import React, { memo } from 'react'
import { ThemeProvider } from 'react-jss'
import { SoraThemeProps } from './SoraTheme.types'

/******************************************************************************
 * Wraps children elements and styles them with the theme passed as prop.
 * @param {ReactNode} children JSX elements to be styled with context.
 * @param {Theme} theme object containing the styles.
 * @returns {React.FC} Wrapper with theme containing styles as a Context.
 *****************************************************************************/
const SoraTheme = memo(function SoraTheme({
  children,
  theme
}: SoraThemeProps) {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
})

export default SoraTheme
