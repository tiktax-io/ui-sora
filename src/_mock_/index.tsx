import { InputDateProps } from '../input-date/InputDate.types'
import { InputPasswordProps } from '../input-password/InputPassword.types'
import { InputSelectProps } from '../input-select/InputSelect.types'
import { InputTextProps } from '../input-text/InputText.types'

export const mockTheme = {
  default_container: {
    background: 'lime',
    border: 'green',
    borderRadius: '4px',
    padding: '20px'
  },
  default_label: {
    color: 'white',
    fontSize: '20px'
  }
}

export const mockInputDateProps: InputDateProps = {
  label: 'Custom label',
  onChange: (newValue: string) => console.log(`New value is: ${newValue}`),
  value: '2025-12-25'
}

export const mockInputPasswordProps: InputPasswordProps = {
  label: 'Custom label',
  onChange: (newValue?: string) => console.log(`New value is: ${newValue}`),
  placeholder: 'Custom placeholder'
}

export const mockInputSelectProps: InputSelectProps = {
  label: 'Custom label',
  onChange: (newValue: string) => console.log(`New value is: ${newValue}`),
  options: [
    'Bitcoin',
    'Ethereum',
    'Bit2me'
  ]
}

export const mockInputTextProps: InputTextProps = {
  label: 'Custom label',
  onChange: (newValue: string) => console.log(`New value is: ${newValue}`),
  placeholder: 'Custom placeholder'
}
