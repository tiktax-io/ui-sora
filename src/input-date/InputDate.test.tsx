import React, { ChangeEvent, useState } from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import InputDate from './InputDate'
import { InputDateProps } from './InputDate.types'
import renderer from 'react-test-renderer'
import { mockTheme } from '../_mock_'
import { ThemeProvider } from 'react-jss'
import SoraTheme from '../sora-theme/SoraTheme'
import darkDefault from '../_styles/official_themes/dark_default'

describe('InputDate component', () => {

  test('Can be queried by \'dataTestId\' prop', () => {
    const props: InputDateProps = {
      dataTestId: 'foundByTestId',
      onChange: () => null
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.queryByTestId('foundByTestId')
    expect(element).toBeInTheDocument()
  })

  test('Accepts value via \'value\' prop', () => {
    const props: InputDateProps = {
      dataTestId: 'valueTest',
      onChange: () => null,
      value: '1990-01-19'
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByTestId('valueTest')
    expect(element).toHaveValue('1990-01-19')
  })

  test('Displays error feedback message if specified', () => {
    const props: InputDateProps = {
      dataTestId: 'withFeedback',
      feedbackMessage: 'Invalid message',
      isInvalid: true,
      onChange: () => null
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByTestId('withFeedbackSpanFeedbackMessage')
    expect(element).toHaveTextContent(/Invalid message/i)
  })

  test('Displays optional error status styled as expected', () => {
    const props: InputDateProps = {
      feedbackMessage: 'Invalid date',
      isInvalid: true,
      onChange: () => null
    }
    const tree = renderer
    .create(
      <InputDate {...props} />
    )
    .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Displays success feedback message if specified', () => {
    const props: InputDateProps = {
      dataTestId: 'successFeedback',
      feedbackMessage: 'Success message',
      isValid: true,
      onChange: () => null
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByTestId('successFeedbackSpanFeedbackMessage')
    expect(element).toHaveTextContent(/Success message/i)
  })

  test('Displays optional success status styled as expected', () => {
    const props: InputDateProps = {
      feedbackMessage: 'Success date',
      isValid: true,
      onChange: () => null
    }
    const tree = renderer
      .create(
        <InputDate {...props} />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('A label can be specified via \'label\' prop.', () => {
    const props: InputDateProps = {
      label: 'My custom label',
      onChange: () => null
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByText(/My custom label/i)
    expect(element).toBeInTheDocument()
  })

  test('Input with labels matched expected styles', () => {
    const props: InputDateProps = {
      label: 'My custom label',
      onChange: () => null
    }
    const tree = renderer
      .create(
        <InputDate {...props} />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Accepts prop \'min\' and pass it to HTML element', () => {
    const props: InputDateProps = {
      dataTestId: 'acceptsMin',
      min: '1990-01-19',
      onChange: () => null,
      value: '1990-01-19'
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByTestId('acceptsMin')
    expect(element).toHaveAttribute('min', '1990-01-19')
  })

  test('Accepts prop \'max\' and pass it to HTML element', () => {
    const props: InputDateProps = {
      dataTestId: 'acceptsMax',
      max: '1990-01-19',
      onChange: () => null,
      value: '1990-01-19'
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByTestId('acceptsMax')
    expect(element).toHaveAttribute('max', '1990-01-19')
  })

  test('Accepts prop \'step\' and pass it to HTML element', () => {
    const props: InputDateProps = {
      dataTestId: 'acceptsStep',
      step: 7,
      onChange: () => null
    }
    render(
      <InputDate {...props} />
    )
    const element = screen.getByTestId('acceptsStep')
    expect(element).toHaveAttribute('step', '7')
  })

  test('On date change triggers function passed as prop \'onChange\'', () => {
    const Container = () => {
      const [value, setValue] = useState<string>('1990-01-30')
      const props: InputDateProps = {
        dataTestId: 'canChange',
        onChange: ({ target }: ChangeEvent<HTMLSelectElement>) =>
          setValue(target?.value),
        value: value
      }
      return (
        <InputDate {...props} />
      )
    }
    render(
      <Container/>
    )
    const element = screen.getByTestId('canChange')
    expect(element).toHaveValue('1990-01-30')
    fireEvent.change(element, {target: {value: '1990-01-19'}})
    expect(element).toHaveValue('1990-01-19')
    fireEvent.change(element, {target: {value: '1990-01-18'}})
    expect(element).toHaveValue('1990-01-18')
  })

  test('Wrapper div be customised via props named \"css\".', () => {
    const props: InputDateProps = {
      css: {
        background: 'red',
        color: 'yellow'
      },
      onChange: () => null,
      value: '1990-01-19'
    }
    const tree = renderer
      .create(
        <InputDate {...props} />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Styles are customized if wrapped by SoraTheme', () => {
    const props: InputDateProps = {
      onChange: () => null
    }
    render(
      <div style={{ alignItems: 'center', display: 'flex', gap: '20px' }}>
        <SoraTheme theme={darkDefault}>
          <InputDate dataTestId='customized' {...props} />
        </SoraTheme>
        <InputDate dataTestId='default1' {...props} />
        <InputDate dataTestId='default2' {...props} />
      </div>
    )
    const CustomInputDate = screen.getByTestId('customized').getAttribute('class')
    const Default1 = screen.getByTestId('default1').getAttribute('class')
    const Default2 = screen.getByTestId('default2').getAttribute('class')
    expect(Default1).toMatch(`${Default2}`)
    expect(CustomInputDate).not.toMatch(`${Default1}`)
    expect(CustomInputDate).not.toMatch(`${Default2}`)
  })

  test('Customized via SoraTheme matches snapshot', () => {
    const props: InputDateProps = {
      onChange: () => null
    }
    const tree = renderer
      .create(
        <SoraTheme theme={darkDefault}>
          <InputDate {...props} />
        </SoraTheme>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

})
