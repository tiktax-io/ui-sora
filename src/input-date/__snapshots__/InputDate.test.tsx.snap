// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDate component Customized via SoraTheme matches snapshot 1`] = `
<div
  className="container-0-2-6"
>
  <input
    className="inputBox-0-2-9"
    onChange={[Function]}
    placeholder="hey"
    style={Object {}}
    type="date"
  />
</div>
`;

exports[`InputDate component Displays optional error status styled as expected 1`] = `
<div
  className="container-0-2-1"
>
  <input
    className="inputBox-0-2-4"
    onChange={[Function]}
    placeholder="hey"
    style={
      Object {
        "border": "1px solid #F45B69",
        "boxShadow": "0px 4px 10px 3px rgba(244, 91, 105, 0.25)",
      }
    }
    type="date"
  />
  <span
    className="feedbackMessageContainer-0-2-2"
    data-testid="undefinedSpanFeedbackMessage"
    style={
      Object {
        "color": "#F45B69",
      }
    }
  >
    <svg
      fill="none"
      height={16}
      viewBox="0 0 16 16"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 0C3.58125 0 0 3.58125 0 8C0 12.4187 3.58125 16 8 16C12.4187 16 16 12.4187 16 8C16 3.58125 12.4187 0 8 0ZM7.25 4.75C7.25 4.3375 7.5875 4 8 4C8.4125 4 8.75 4.33594 8.75 4.75V8.75C8.75 9.16406 8.41406 9.5 8 9.5C7.58594 9.5 7.25 9.16563 7.25 8.75V4.75ZM8 12.5C7.4575 12.5 7.0175 12.06 7.0175 11.5175C7.0175 10.975 7.45719 10.535 8 10.535C8.54281 10.535 8.9825 10.975 8.9825 11.5175C8.98125 12.0594 8.54375 12.5 8 12.5Z"
        fill="#F45B69"
      />
    </svg>
    <span
      className="feedbackMessageText-0-2-3"
    >
      Invalid date
    </span>
  </span>
</div>
`;

exports[`InputDate component Displays optional success status styled as expected 1`] = `
<div
  className="container-0-2-1"
>
  <input
    className="inputBox-0-2-4"
    onChange={[Function]}
    placeholder="hey"
    style={
      Object {
        "border": "1px solid #7CEA9C",
        "boxShadow": "0px 4px 10px 3px rgba(124, 234, 156, 0.25)",
      }
    }
    type="date"
  />
  <span
    className="feedbackMessageContainer-0-2-2"
    data-testid="undefinedSpanFeedbackMessage"
    style={
      Object {
        "color": "#7CEA9C",
      }
    }
  >
    <svg
      fill="none"
      height={16}
      viewBox="0 0 16 16"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 8C0 3.58125 3.58125 0 8 0C12.4187 0 16 3.58125 16 8C16 12.4187 12.4187 16 8 16C3.58125 16 0 12.4187 0 8ZM11.6187 6.61875C11.9594 6.27813 11.9594 5.72187 11.6187 5.38125C11.2781 5.04062 10.7219 5.04062 10.3813 5.38125L7 8.7625L5.61875 7.38125C5.27813 7.04062 4.72187 7.04062 4.38125 7.38125C4.04062 7.72187 4.04062 8.27812 4.38125 8.61875L6.38125 10.6187C6.72187 10.9594 7.27813 10.9594 7.61875 10.6187L11.6187 6.61875Z"
        fill="#7CEA9C"
      />
    </svg>
    <span
      className="feedbackMessageText-0-2-3"
    >
      Success date
    </span>
  </span>
</div>
`;

exports[`InputDate component Input with labels matched expected styles 1`] = `
<div
  className="container-0-2-1"
>
  <label
    className="label-0-2-5"
  >
    My custom label
  </label>
  <input
    className="inputBox-0-2-4"
    onChange={[Function]}
    placeholder="hey"
    style={Object {}}
    type="date"
  />
</div>
`;

exports[`InputDate component Wrapper div be customised via props named "css". 1`] = `
<div
  className="container-0-2-1"
>
  <input
    className="inputBox-0-2-4"
    onChange={[Function]}
    placeholder="hey"
    style={
      Object {
        "background": "red",
        "color": "yellow",
      }
    }
    type="date"
    value="1990-01-19"
  />
</div>
`;
